controllers/cassandrabackup_controller.go

```diff
@@ -132,7 +191,9 @@ func (r *CassandraBackupReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 	}
 
 	for _, p := range pods {
-		go func(pod corev1.Pod) {
+		pod := p
+		backup := backup.DeepCopy()
```

Comment:
It doesn't look like this completely fixed the race condition.

```
 kubectl get cassandrabackup backup-1 -o yaml
...
status:
  failed:
  - test-dc2-default-sts-1
  finished:
  - test-dc2-default-sts-2
  inProgress:
  - test-dc2-default-sts-0
  - test-dc2-default-sts-2
  - test-dc2-default-sts-2
  - test-dc2-default-sts-0
  startTime: "2021-03-25T21:02:38Z
```
test-dc2-default-sts-1 did in fact fail. Both test-dc2-default-sts-2 and test-dc2-default-sts-1 completed the operation successfully though.

Reply:
I'll continue tomorrow, but at least this one makes the InProgress test go through correctly.

```diff
diff --git a/controllers/cassandrabackup_controller.go b/controllers/cassandrabackup_controller.go
index cb76148..78a54d7 100644
--- a/controllers/cassandrabackup_controller.go
+++ b/controllers/cassandrabackup_controller.go
@@ -19,6 +19,7 @@ package controllers
 import (
        "context"
        "fmt"
+       "sync"
        "time"
 
        "github.com/go-logr/logr"
@@ -163,11 +164,6 @@ func (r *CassandraBackupReconciler) createBackup(ctx context.Context, backup *ap
                return ctrl.Result{RequeueAfter: 10 * time.Second}, err
        }
 
-       if err = r.addCassdcSpecToStatus(ctx, backup, cassdc); err != nil {
-               r.Log.Error(err, "failed to patch status with CassdcTemplateSpec", "CassandraDatacenter", cassdcKey)
-               return ctrl.Result{RequeueAfter: 10 * time.Second}, err
-       }
-
        pods, err := r.getCassandraDatacenterPods(ctx, cassdc)
        if err != nil {
                return ctrl.Result{RequeueAfter: 10 * time.Second}, err
@@ -181,6 +177,11 @@ func (r *CassandraBackupReconciler) createBackup(ctx context.Context, backup *ap
        }
 
        patch := client.MergeFrom(backup.DeepCopy())
+       if err = r.addCassdcSpecToStatus(ctx, backup, cassdc); err != nil {
+               r.Log.Error(err, "failed to patch status with CassdcTemplateSpec", "CassandraDatacenter", cassdcKey)
+               return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+       }
+
        backup.Status.StartTime = metav1.Now()
        for _, pod := range pods {
                backup.Status.InProgress = append(backup.Status.InProgress, pod.Name)
@@ -190,9 +191,12 @@ func (r *CassandraBackupReconciler) createBackup(ctx context.Context, backup *ap
                return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, err
        }
 
+       backupMutex := sync.Mutex{}
+       patch = client.MergeFrom(backup.DeepCopy())
+       wg := sync.WaitGroup{}
        for _, p := range pods {
                pod := p
-               backup := backup.DeepCopy()
+               wg.Add(1)
                go func() {
                        r.Log.Info("starting backup", "CassandraPod", pod.Name)
                        succeeded := false
@@ -202,18 +206,21 @@ func (r *CassandraBackupReconciler) createBackup(ctx context.Context, backup *ap
                        } else {
                                r.Log.Error(err, "backup failed", "CassandraPod", pod.Name)
                        }
-                       patch := client.MergeFrom(backup.DeepCopy())
+                       backupMutex.Lock()
+                       defer backupMutex.Unlock()
+                       defer wg.Done()
                        backup.Status.InProgress = removeValue(backup.Status.InProgress, pod.Name)
                        if succeeded {
                                backup.Status.Finished = append(backup.Status.Finished, pod.Name)
                        } else {
                                backup.Status.Failed = append(backup.Status.Failed, pod.Name)
                        }
-                       if err := r.Status().Patch(context.Background(), backup, patch); err != nil {
-                               r.Log.Error(err, "failed to patch status", "Backup", backup)
-                       }
                }()
        }
+       wg.Wait()
+       if err := r.Status().Patch(context.Background(), backup, patch); err != nil {
+               r.Log.Error(err, "failed to patch status", "Backup", backup)
+       }
        return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 }
 
@@ -295,10 +302,11 @@ func (r *CassandraBackupReconciler) addCassdcSpecToStatus(ctx context.Context, b
                },
        }
 
-       patch := client.MergeFrom(backup.DeepCopy())
+       // patch := client.MergeFrom(backup.DeepCopy())
        backup.Status.CassdcTemplateSpec = templateSpec
 
-       return r.Status().Patch(ctx, backup, patch)
+       // return r.Status().Patch(ctx, backup, patch)
+       return nil
 }
 
 func (r *CassandraBackupReconciler) getCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.CassandraDatacenter) ([]corev1.Pod, error) {
diff --git a/controllers/cassandrabackup_controller_test.go b/controllers/cassandrabackup_controller_test.go
index 8baab99..9727508 100644
--- a/controllers/cassandrabackup_controller_test.go
+++ b/controllers/cassandrabackup_controller_test.go
@@ -204,7 +204,7 @@ var _ = Describe("CassandraBackup controller", func() {
                        if err != nil {
                                return false
                        }
-                       return len(updated.Status.Finished) > 0
+                       return len(updated.Status.Finished) == 3 && len(updated.Status.InProgress) == 0
                }, timeout, interval).Should(BeTrue())
 
                By("verify that medusa gRPC clients are invoked")
```